lib/clientv2/resources.go

```diff
@@ -0,0 +1,281 @@
+			ErroredFields: []cerrors.ErroredField{{
+				Name:   "Metadata.ResourceVersion",
+				Reason: "field must be set for an Update request",
+				Value:  in.GetObjectMeta().GetResourceVersion(),
+			}},
+		}
+	}
+
+	// Handle namespace field processing common to Create and Update.
+	if err := c.handleNamespace(ns, kind, in); err != nil {
+		return nil, err
+	}
+
+	// Convert the resource to a KVPair and pass that to the backend datastore, converting
+	// the response (if we get one) back to a resource.
+	kvp, err := c.backend.Update(c.resourceToKVPair(opts, kind, in))
+	if kvp != nil {
+		return c.kvPairToResource(kvp), err
+	}
+	return nil, err
+}
+
+// Delete deletes a resource from the backend datastore.
+func (c *resources) Delete(opts options.DeleteOptions, kind, ns, name string) error {
+	// Create a ResourceKey and pass that to the backend datastore.
+	key := model.ResourceKey{
+		Kind:      kind,
+		Name:      name,
+		Namespace: ns,
+	}
+	return c.backend.Delete(key, opts.ResourceVersion)
+}
+
+// Get gets a resource from the backend datastore.
+func (c *resources) Get(opts options.GetOptions, kind, ns, name string) (resource, error) {
+	key := model.ResourceKey{
+		Kind:      kind,
+		Name:      name,
+		Namespace: ns,
+	}
+	kvp, err := c.backend.Get(key, opts.ResourceVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := c.kvPairToResource(kvp)
+	return out, nil
+}
+
+// List lists a resource from the backend datastore.
+func (c *resources) List(opts options.ListOptions, kind, listKind, ns, name string, listObj resourceList) error {
+	key := model.ResourceListOptions{
+		Kind:      kind,
+		Name:      name,
+		Namespace: ns,
+	}
+
+	// Query the backend.
+	kvps, err := c.backend.List(key, opts.ResourceVersion)
+	if err != nil {
+		return err
+	}
+
+	// Convert the slice of KVPairs to a slice of Objects.
+	resources := []runtime.Object{}
+	for _, kvp := range kvps.KVPairs {
+		resources = append(resources, c.kvPairToResource(kvp))
+	}
+	err = meta.SetList(listObj, resources)
+	if err != nil {
+		return err
+	}
+
+	// Finally, set the resource version and api group version of the list object.
+	listObj.GetListMeta().SetResourceVersion(kvps.Revision)
+	listObj.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   apiv2.Group,
+		Version: apiv2.VersionCurrent,
+		Kind:    listKind,
+	})
+
+	return nil
+}
+
+// Watch watches a specific resource or resource type.
+func (c *resources) Watch(opts options.ListOptions, kind, ns, name string) (watch.Interface, error) {
+	panic("Not implemented")
+	return nil, nil
+}
+
+// resourceToKVPair converts the resource to a KVPair that can be consumed by the
+// backend datastore client.
+func (c *resources) resourceToKVPair(opts options.SetOptions, kind string, in resource) *model.KVPair {
+	// Prepare the resource to remove non-persisted fields.
+	rv := in.GetObjectMeta().GetResourceVersion()
+	in.GetObjectMeta().SetResourceVersion("")
+	in.GetObjectMeta().SetSelfLink("")
```

Comment:
what about other things in the ObjectMetadata like `creationTimestamp`, and `UID`?

Reply:
those things are persisted in the datastore so should not be removed.